packages/mdc-tab-scroller/foundation.js

```diff
@@ -0,0 +1,111 @@
+/**
+  * @license
+  * Copyright 2018 Google Inc. All Rights Reserved.
+  *
+  * Licensed under the Apache License, Version 2.0 (the "License")
+  * you may not use this file except in compliance with the License.
+  * You may obtain a copy of the License at
+  *
+  *      http://www.apache.org/licenses/LICENSE-2.0
+  *
+  * Unless required by applicable law or agreed to in writing, software
+  * distributed under the License is distributed on an "AS IS" BASIS,
+  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  * See the License for the specific language governing permissions and
+  * limitations under the License.
+*/
+
+import MDCFoundation from '@material/base/foundation';
+import MDCTabScrollerAdapter from './adapter';
+import {
+  cssClasses,
+  strings,
+} from './constants';
+
+/**
+ * @extends {MDCFoundation<!MDCTabScrollerAdapter>}
+ * @abstract
+ */
+class MDCTabScrollerFoundation extends MDCFoundation {
+  /** @return enum {string} */
+  static get cssClasses() {
+    return cssClasses;
+  }
+
+  /** @return enum {string} */
+  static get strings() {
+    return strings;
+  }
+
+  /**
+   * @see MDCTabScrollerAdapter for typing information
+   * @return {!MDCTabScrollerAdapter}
+   */
+  static get defaultAdapter() {
+    return /** @type {!MDCTabScrollerAdapter} */ ({
+      registerEventHandler: () => {},
+      deregisterEventHandler: () => {},
+      addClass: () => {},
+      removeClass: () => {},
+      setContentStyleProperty: () => {},
+      getContentStyleValue: () => {},
+      setScrollLeft: () => {},
+      getScrollLeft: () => {},
+      computeContentClientRect: () => {},
+      computeClientRect: () => {},
+    });
+  }
+
+  /** @param {!MDCTabScrollerAdapter} adapter */
+  constructor(adapter) {
+    super(Object.assign(MDCTabScrollerFoundation.defaultAdapter, adapter));
+  }
+
+  /**
+   * Scrolls to the given scrollX value
+   * @param {number} scrollX
+   * @abstract
+   */
+  scrollTo(scrollX) {} // eslint-disable-line no-unused-vars
+
+  /**
+   * Returns the translateX value from a CSS matrix transform function string
+   * @return {number}
+   * @protected
+   */
+  calculateCurrentTranslateX() {
+    const transformValue = this.adapter_.getContentStyleValue('transform');
+    // Early exit if no transform is present
+    if (transformValue === 'none') {
+      return 0;
+    }
+
+    // The transform value comes back as a matrix transformation in the form
+    // of `matrix(a, b, c, d, tx, ty)`. We only care about tx (translateX) so
+    // we're going to grab all the parenthesizedvalues, strip out tx, and parse it.
+    const results = /\((.+)\)/.exec(transformValue)[1];
+    const parts = results.split(',');
+    return parseFloat(parts[4]);
+  }
+
+  /**
+   * Calculates a safe scroll value that is > 0 and < the max scroll value
+   * @param {number} scrollX The distance to scroll
+   * @return {number}
+   * @protected
+   */
+  calculateSafeScrollValue(scrollX) {
```

Comment:
You could change `scrollTo` to first do this "clamping" logic. and then call an `internalScrollTo` which each subclass.

Actually I'm not sure where calculateSafeScrollValue ever gets called

Reply:
I use it in the concrete foundations.

https://github.com/material-components/material-components-web/pull/2577/files#diff-e008d1c7b64ae7e67f7d2ef61e7e29d7R43

https://github.com/material-components/material-components-web/pull/2577/files#diff-1d7b625beb8cb4f97fd1a08836fe195bR76

Reply:
Okay just sort this protected method below the public methods then.